data/dataGeneratorScripts/generateRenewables/importMarktStammDaten.py

```diff
@@ -0,0 +1,372 @@
+from collections import defaultdict
+import csv
+from lxml import etree
+import os
+import re
+import json
+
+
+cutoffdate="31.12.2018"
+
+# this is only used in the handling of "not 2018"-ags keys 
+# this is not used in the xml parsing so far
+
+exceptionDict={
+"09478444":"Neuensorger Forst ist ein gemeindefreies gebiet",
+"01053105":"Sachsenwald ist ein gemeindefreies Gebiet",
+"06435200":"Gutsbezirk Spessart ist ist ein gemeindefreies Gebiet",
+"01060014":"BuchholzerForstist ist ein gemeindefreies Gebiet",
+"09572444":"Gdefr. Geb. (Lkr Erlangen-Höchstadt) ist ein gemeindefreies Gebiet",
+"03155501":"Solling (Landkreis Northeim) ist ein gemeindefreies Gebiet",
+"03153504":"Harz (Landkreis Goslar), gemfr. Gebiet ist ein gemeindefreies Gebiet"}
+
+
+
+def parseElem(elem,tags:list[str]) -> list:
+    returnList = [None]*len(tags)
+    for subelem in elem:
+        print(subelem.tag+"   "+subelem.text)
+        for i,tag in enumerate(tags):
+            if subelem.tag == tag:
+                returnList[i] = subelem.text           
+    return returnList
+        
+def parseXML(path:str,*, useinputDict = False, inputDict = defaultdict(float) ,printInfo:bool=False, printExit:bool=False) -> defaultdict:
+    """This parses the XML file under path and returns a dict formated list {key: power | ags keys in xml}
+    Right now we also filter out some entries due to some specifications like the installation date.
+    There are just a few entries (<10) in the xml that do not ags keys. ATM we disregard those.  
+    """
+    returnDict = defaultdict(float)
+    if useinputDict:
+        returnDict=inputDict
+    with open(path,"r", encoding='utf-16') as file: 
+
+        tree = etree.parse(file, parser = etree.XMLParser())
+        root = tree.getroot()
+
+        noAgsCount = 0
+        noPowerCount = 0
+        noDate = 0 
+        for elem in root:           
+            [ags,netpower,date,enddate,status] = parseElem(elem,["Gemeindeschluessel","Nettonennleistung","Inbetriebnahmedatum","DatumEndgueltigeStilllegung","EinheitBetriebsstatus"])
+            
+            #Betriebsstatus
+            # 35 -> In Betrieb, 31 -> In Planung, 37 -> Vorübergehend stillgelegt, 38 -> Endgültig stillgelegt
+            if status == "31": #filter out "in Planung"
+                continue
+
+            if enddate is not None and int(enddate[:4]) < 2019:    #filter out "stillgelegt vor 2019"
+                continue                
```

Comment:
One check that I believe is missing here whether a unit was temporarily out of order on 2018-12-31. There are a few:

```
sqlite3 --readonly --table --header Marktstammdatenregister.db <temp.sql 
+--------------------+--------------------+---------------------+---------------------+-----------------------------+---------------------------------------+----------------------------+
| EinheitMastrNummer | Gemeindeschluessel | Registrierungsdatum | Inbetriebnahmedatum | DatumEndgueltigeStilllegung | DatumBeginnVoruebergehendeStilllegung | DatumWiederaufnahmeBetrieb |
+--------------------+--------------------+---------------------+---------------------+-----------------------------+---------------------------------------+----------------------------+
| SEE940917235426    | 06431003           | 2021-06-11          | 1992-07-13          |                             | 2013-01-01                            |                            |
| SEE970950685179    | 09471208           | 2021-02-22          | 1998-01-01          |                             | 2018-07-01                            |                            |
| SEE908579456740    | 14521180           | 2019-12-30          | 2000-11-27          |                             | 2018-05-28                            |                            |
| SEE946498391087    | 06435021           | 2021-06-21          | 2001-05-09          | 2016-12-03                  | 2016-12-02                            |                            |
| SEE924641851886    | 08435057           | 2021-03-02          | 2001-12-01          |                             | 2013-10-15                            |                            |
| SEE963007354166    | 08222000           | 2020-08-05          | 2002-12-17          |                             | 2015-08-03                            | 2020-08-04                 |
| SEE953246223538    | 08222000           | 2020-08-04          | 2004-07-26          |                             | 2015-08-03                            |                            |
| SEE984820500431    | 08222000           | 2020-08-04          | 2004-07-26          |                             | 2015-08-04                            |                            |
| SEE985573680722    | 08222000           | 2020-08-18          | 2004-07-26          |                             | 2015-01-01                            |                            |
| SEE988733115041    | 06635018           | 2019-03-02          | 2006-09-08          | 2016-01-01                  | 2015-12-31                            |                            |
| SEE936191315770    | 08119018           | 2020-09-23          | 2006-12-15          |                             | 2015-02-23                            | 2020-05-09                 |
| SEE933515647708    | 03360001           | 2021-02-07          | 2008-04-30          |                             | 2016-12-12                            |                            |
| SEE907763705213    | 09180117           | 2020-11-30          | 2009-05-19          |                             | 2017-11-27                            |                            |
| SEE995170337905    | 05911000           | 2021-01-28          | 2009-11-05          |                             | 2018-03-12                            |                            |
| SEE934689600001    | 05154036           | 2021-01-31          | 2010-06-01          |                             | 2018-12-28                            |                            |
| SEE937282159198    | 03360001           | 2021-02-07          | 2010-06-30          |                             | 2016-12-12                            |                            |
| SEE929943807784    | 08435057           | 2021-03-02          | 2010-12-13          |                             | 2013-10-15                            |                            |
| SEE986230459743    | 13071096           | 2020-01-14          | 2011-09-30          |                             | 2012-10-06                            |                            |
| SEE947572090827    | 08336014           | 2020-08-11          | 2011-12-15          |                             | 2011-12-16                            |                            |
| SEE914810546941    | 08326052           | 2019-05-23          | 2011-12-19          |                             | 2011-12-20                            |                            |
| SEE987563779416    | 05758004           | 2020-11-30          | 2011-12-29          |                             | 2013-12-29                            |                            |
| SEE974929724524    | 08426013           | 2021-09-03          | 2011-12-31          |                             | 2018-08-01                            | 2021-02-24                 |
| SEE966845109103    | 02000000           | 2019-08-09          | 2012-03-05          |                             | 2018-02-16                            |                            |
| SEE948748543444    | 16051000           | 2021-01-31          | 2012-06-05          |                             | 2016-12-13                            |                            |
| SEE933708173960    | 07315000           | 2021-02-12          | 2013-03-28          |                             | 2017-12-06                            |                            |
| SEE942263936569    | 07337070           | 2020-07-05          | 2014-07-31          |                             | 2018-01-01                            | 2020-06-24                 |
| SEE959082082117    | 06440017           | 2021-02-11          | 2014-10-31          |                             | 2018-07-25                            | 2020-12-07                 |
| SEE981269470325    | 01057077           | 2019-03-03          | 2017-05-15          | 2017-05-17                  | 2017-05-16                            |                            |
| SEE964966236203    | 05112000           | 2020-12-31          | 2017-09-18          |                             | 2017-09-24                            | 2019-06-30                 |
| SEE921772502595    | 05112000           | 2020-12-31          | 2017-09-25          |                             | 2017-10-04                            |                            |
| SEE967773395460    | 12063142           | 2019-03-14          | 2018-12-08          |                             | 2018-12-09                            | 2019-03-14                 |
+--------------------+--------------------+---------------------+---------------------+-----------------------------+---------------------------------------+----------------------------+
```

Generated with

```sql
select
  EinheitMastrNummer,
  Gemeindeschluessel,
  Registrierungsdatum,
  Inbetriebnahmedatum,
  DatumEndgueltigeStilllegung,
  DatumBeginnVoruebergehendeStilllegung,
  DatumWiederaufnahmeBetrieb
from
  EinheitSolar
where
  Inbetriebnahmedatum <= "2018-12-31"
  and DatumBeginnVoruebergehendeStilllegung <= "2018-12-31"
  and (DatumWiederaufnahmeBetrieb is null or DatumWiederaufnahmeBetrieb > "2018-12-31")
```

To be fair, their total Nettonennleistung is only 368.985 kW.

[temp.sql](https://github.com/GermanZero-de/localzero-generator-core/files/8399330/temp.sql.txt)
How to obtain DB: https://github.com/GermanZero-de/localzero-data-public/issues/29

Reply:
Ja das stimmt. Das sollte wahrscheinlich noch hinzugefügt werden. Obwohl 369 kW eigentlich vernachlässigbar über die ~30 kommunen ist.

Aber füge ich noch hinzu! Danke für das Feedback :)